lib/iris/experimental/ugrid.py

```diff
@@ -2903,31 +2903,70 @@ def _construct_access_arrays(self):
             the result is always based on current content in the Mesh.
 
         """
-        # TODO: cheat for now : correct calculations, but not dynamic.
-        # TODO: replace wth fully dynamic lazy calcs (slightly more complex).
         mesh, location, axis = self.mesh, self.location, self.axis
         node_coord = self.mesh.coord(include_nodes=True, axis=axis)
+
         if location == "node":
-            points = node_coord.core_points()
-            bounds = None
+            points_coord = node_coord
+            bounds_connectivity = None
         elif location == "edge":
-            edge_coord = self.mesh.coord(include_edges=True, axis=self.axis)
-            points = edge_coord.core_points()
-            inds = mesh.edge_node_connectivity.core_indices()
-            bounds = node_coord.core_points()[inds]
+            points_coord = self.mesh.coord(include_edges=True, axis=axis)
+            bounds_connectivity = mesh.edge_node_connectivity
         elif location == "face":
-            face_coord = self.mesh.coord(include_faces=True, axis=self.axis)
-            points = face_coord.core_points()  # For now, this always exists
-            inds = mesh.face_node_connectivity.core_indices()
-            bounds = node_coord.core_points()[inds]
+            points_coord = self.mesh.coord(include_faces=True, axis=axis)
+            bounds_connectivity = mesh.face_node_connectivity
+
+        # The points output is the points of the relevant element-type coord.
+        points = points_coord.core_points()
+        if bounds_connectivity is None:
+            bounds = None
+        else:
+            # Bounds are calculated from a connectivity and the node points.
+            # Data can be real or lazy, so operations must work in Dask, too.
+            indices = bounds_connectivity.core_indices()
+            # Normalise indices dimension order to [faces/edges, bounds]
+            indices = bounds_connectivity.indices_by_src(indices)
+            # Normalise the start index
+            indices = indices - bounds_connectivity.start_index
+
+            node_points = node_coord.core_points()
+            n_nodes = node_points.shape[0]
+            # Choose real/lazy array library, to suit array types.
+            lazy = _lazy.is_lazy_data(indices) or _lazy.is_lazy_data(
+                node_points
+            )
+            al = da if lazy else np
+            # NOTE: Dask cannot index with a multidimensional array, so we
+            # must flatten it and restore the shape later.
+            flat_inds = indices.flatten()
+            # NOTE: the connectivity array can have masked points, but we can't
+            # effectively index with those.  So use a non-masked index array
+            # with "safe" index values, and post-mask the results.
+            flat_inds_mask = al.ma.getmaskarray(flat_inds)
+            flat_inds_nomask = al.ma.filled(flat_inds, 0)
```

Comment:
hi @bjlittle
Don't want to interrupt again if you're already looking into this,
but TBH I think this can still be improved.

I previously was tripped up + made it more complicated because I found "True OR masked --> masked".

So I **_think_** we can have here :
```
flat_inds_nomask = al.ma.filled(flat_inds, -1)
missing_inds = (flat_inds_nomask < 0) | (flat_inds_nomask >= n_nodes)
flat_inds_safe = al.where(missing_inds, 0, flat_inds_nomask)
```

So, 
 (1) use "filled" to combine mask with 'invalid' points, and
 (2) "|" operator instead of 'logical_or'.

There is also scope here for a bit more testing : I haven't checked the handling of non-masked out-of-range index values.
Let me know if you're up for this further change.

Reply:
>  this further change.

If you have the appetite, I have some additional including testing ready to add on